File: api/src/Features/Services/Add/Service.cs
using Microsoft.AspNetCore.Mvc;

using Npgsql;

using Kratos.Api.Common.Types;
using Kratos.Api.Common.Constants;
using Kratos.Api.Common.Services;
using Kratos.Api.Common.Extensions;

namespace Kratos.Api.Features.Services.Add;

public class Service(
    [FromServices] IImageUploadService imageUploadService,
    [FromServices] IRepository repo
)
{
    public async Task<Result> AddAsync(Request request, CancellationToken cancellationToken)
    {
        string seoFriendlyServiceName = request.Name.SeoFriendly();

        bool serviceNameExists = await repo.ExistsAsync(seoFriendlyServiceName, cancellationToken);
        if (serviceNameExists)
        {
            return Result.ConflictError("Service name conflicts with another");
        }

        if (request.ParentServiceId is not null)
        {
            bool parentServiceExists = await repo.ExistsAsync(request.ParentServiceId.Value, cancellationToken);
            if (!parentServiceExists)
            {
                return Result.NotFoundError("Parent service not found");
            }
        }

        string imageFileName = await imageUploadService.UploadImageAsync(Folders.Upload.Services, request.Image, cancellationToken);

        try
        {
            Database.Models.Service service = request.AsService(imageFileName, seoFriendlyServiceName);
            await repo.AddServiceAsync(service, cancellationToken);

            return Result.Success();
        }
        catch (PostgresException)
        {
            return Result.ConflictError("Service name conflicts with another");
        }
    }
}

File: client/web/src/features/services/types/service.ts
import type { Column } from "@/types/column";

export type Service = {
  id: number;
  name: string;
  description: string;
  summary: string;
  imageFileName: string;
  parentServiceName?: string;
  parentServiceId?: number;
};

export type AddServiceRequest = {
  name: string;
  description: string;
  summary: string;
  image: File;
  parentServiceId?: number;
};

export type EditServiceRequest = {
  id: number;
  name: string;
  description: string;
  summary: string;
  image?: File;
  parentServiceId?: number;
};

export function createServiceMappings(): Column[] {
  return [
    {
      header: "Name",
      accessorKey: "name",
    },
    {
      header: "Summary",
      accessorKey: "summary",
    },
    {
      header: "Parent Service Name",
      accessorKey: "parentServiceName",
    },
  ];
}
